### PR TITLE
MarlinBook

### PR DESCRIPTION
**Pull request by JBenda**
BEGINRELEASENOTES

- Added Base for **MarlinBook**
   - allow the creation of objects for later easy parallel use
- Added fundamental test for the Module 
- Added DataModel UML (`/docs/diagrams/`)

ENDRELEASENOTES

- [x] check coding convention
- [x] adding exception handleing
- [x] adding titlet Histograms
- [x] replace tabs with spaes

### CODE REVIEW

File: doc/Doxyfile [hunk 1]
```diff
@@ -468,8 +468,7 @@ INPUT                  = $(DOX_INPUT)
 # *.c *.cc *.cxx *.cpp *.c++ *.java *.ii *.ixx *.ipp *.i++ *.inl *.h *.hh *.hxx 
 # *.hpp *.h++ *.idl *.odl *.cs *.php *.php3 *.inc *.m *.mm *.py
 
-FILE_PATTERNS          = *.h \
-                         *.cc
+FILE_PATTERNS          = *.h
```
Comment (JBenda) [inline doc/Doxyfile]: in .cc files are no public documentation, but template spezialisations and Macros, which Doxygen can't handle

File: doc/mkdoxygen/Doxyfile [hunk 2]
```diff
@@ -459,7 +459,7 @@ WARN_LOGFILE           =
 # directories like "/usr/src/myproject". Separate the files or directories
 # with spaces.
 
-INPUT                  = overview.txt ../../source/include ../../source/src ../../source/tinyxml
+INPUT                  = overview.txt ../../source/include ../../source/src ../../source/tinyxml ../../source/book
```
Comment (JBenda) [inline doc/mkdoxygen/Doxyfile]: added MarlinBook to Doxygen

File: doc/mkdoxygen/Doxyfile [hunk 3]
```diff
@@ -468,8 +468,7 @@ INPUT                  = overview.txt ../../source/include ../../source/src ../.
 # *.c *.cc *.cxx *.cpp *.c++ *.java *.ii *.ixx *.ipp *.i++ *.inl *.h *.hh *.hxx
 # *.hpp *.h++ *.idl *.odl *.cs *.php *.php3 *.inc *.m *.mm *.py
 
-FILE_PATTERNS          = *.h \
-                         *.cc
+FILE_PATTERNS          = *.h
```
Comment (JBenda) [inline doc/mkdoxygen/Doxyfile]: same as above
> in .cc files are no public documentation, but template specializations and Macros, which Doxygen can't handle

File: source/include/marlin/Exceptions.h [hunk 4]
```diff
@@ -137,5 +137,6 @@ namespace marlin {
   MARLIN_DEFINE_EXCEPTION( SkipEventException ) ;
   MARLIN_DEFINE_EXCEPTION( StopProcessingException ) ;
   MARLIN_DEFINE_EXCEPTION( ParseException ) ;
+  MARLIN_DEFINE_EXCEPTION( BookStoreException );
```
Comment (JBenda) [inline source/include/marlin/Exceptions.h:140]: Not in use, because of no clear idea of solving double booking.

File: test/tests/test-book-store.cc [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+
+
+int main(int, char**) {
+	marlin::test::UnitTest test(" Memory Filler Test ");
+	BookStore store{};
+	
+	{
+
+		// EntrySingle entry = store.book<RH1F, RAxisConfig>("path", "name", {"a", 3, 1.0, 2.0}) ;	
+		EntrySingle entry = store.book("path", "name", EntryData<RH1F>({"a", 3, 1.0, 2.0}).single());
```
Comment (JBenda) [inline test/tests/test-book-store.cc]: the two ways of booking, prefere 2.

File: doc/Doxyfile (hunk 1 above)
Comment (rete) [reply doc/Doxyfile]: No add it again. The source code is appended after the class documentation normally. If not, it should be and we should fix it then

File: doc/mkdoxygen/Doxyfile (hunk 3 above)
Comment (rete) [reply doc/mkdoxygen/Doxyfile]: Same as above too

File: mkdocs.yml [hunk 6]
```diff
@@ -14,6 +14,7 @@ nav:
       - XML parser: 'user-manual/xml-parser.md'
       - Processor API: 'user-manual/processor-api.md'
       - Multi-threading with MarlinMT: 'user-manual/marlin-mt.md'
+      - Use global Objects in Parallel: 'user-manual/booking-system.md'
```
Comment (rete) [inline mkdocs.yml]: `Use global Objects in Parallel` -> `Histograms booking`

File: source/CMakeLists.txt [hunk 7]
```diff
@@ -34,8 +35,12 @@ ENDIF()
 IF( NOT MARLIN_GEAR )
     LIST( REMOVE_ITEM library_sources ./src/plugins/GearGeometry.cc )
 ENDIF()
+IF( MARLIN_BOOK )
+	# LIST(…) TODO when in framework
```
Comment (rete) [inline source/CMakeLists.txt]: Still TODO ?

File: .gitignore [hunk 8]
```diff
@@ -10,4 +10,5 @@
 /include/
 /lib/
 /*~
-/source/include/marlin/MarlinConfig.h
```
Comment (rete) [inline .gitignore:13]: Why did you remove it ?

File: CMakeLists.txt [hunk 9]
```diff
@@ -141,7 +141,7 @@ INSTALL( FILES "${PROJECT_SOURCE_DIR}/source/python/compareMarlinSteeringFiles.p
 	      OWNER_WRITE )
 # -----------------------------------------------------------------------------
 
-
+set(MARLIN_BOOK ${MARLIN_BOOK} CACHE BOOL "not again" FORCE)
```
Comment (rete) [inline CMakeLists.txt]: "not again" ?

File: source/book/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,19 @@
+#################################
+# cmake file for building MarlinBOOK component
+# @author Remi Ete, DESY
```
Comment (rete) [inline source/book/CMakeLists.txt]: Put your name their too

File: source/book/include/marlin/book/Entry.h [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    template < typename T >
+    class Handle ;
+
+    /**
+     *	@brief trait for type Specific Data, Function and types.
+     */
+    template < typename T >
+    struct trait {
+
+      /// Function to merge two objects of type T.
```
Comment (rete) [inline source/book/include/marlin/book/Entry.h]: Do you need this trait class ?

File: source/book/include/marlin/book/Entry.h [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+
+        } else if ( _key.flags.Contains( Flags::Book::MultiCopy ) ) {
+          return std::static_pointer_cast< EntryMultiCopy< T > >( _entry )
+            ->handle( idx ) ;
+
+        } else if ( _key.flags.Contains( Flags::Book::MultiShared ) ) {
+          return std::static_pointer_cast< EntryMultiShared< T > >( _entry )
+            ->handle() ;
+        }
+        // TODO add log
```
Comment (rete) [inline source/book/include/marlin/book/Entry.h]: TODO ?

File: source/book/include/marlin/book/EntryData.h [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+      EntryKey() : type{std::type_index( typeid( void ) )} {}
+
+      // TODO: clean solution
+      /// Construct typed EntryKey.
+      EntryKey( const std::type_index &t ) : type{t} {}
+
+      /// virtual Entry path
+      std::string path{""},
+        /// Entry name
+        name{""} ;
```
Comment (rete) [inline source/book/include/marlin/book/EntryData.h]: Declare the type here on a new line

File: source/book/include/marlin/book/EntryData.h [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+      /// number of memory instances
+      std::size_t amt{0} ;
```
Comment (rete) [inline source/book/include/marlin/book/EntryData.h]: `amt` ? `nInstances` maybe ?

File: source/book/include/marlin/book/MemLayout.h [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+      template < typename T >
+      std::shared_ptr< const T > merged() {
+        return std::static_pointer_cast< const T >( imp_merged() ) ;
+      }
+
+      virtual ~MemLayout() = default ;
+
+    protected:
+      /// implementation from at
+      virtual std::shared_ptr< void > imp_at( std::size_t idx ) const = 0 ;
```
Comment (rete) [inline source/book/include/marlin/book/MemLayout.h]: `impAt` to cope with conventions

File: source/book/include/marlin/book/MemLayout.h [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+      /// implementation from merged
+      virtual std::shared_ptr< void > imp_merged() = 0 ;
```
Comment (rete) [inline source/book/include/marlin/book/MemLayout.h]: `impMerged` to cope with conventions

File: source/book/include/marlin/book/ROOTAdapter.h [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  namespace book {
+    /// Alias for Types used by MarlinBook
+    namespace types {
+
+      /// Type for Describing a Histogram Axis.
+      using RAxisConfig = ROOT::Experimental::RAxisConfig ;
+
+      /// Generic Histogram.
+      template < int D, typename T, template < int, class > class... STAT >
+      using RH = ROOT::Experimental::RHist< D, T, STAT... > ;
```
Comment (rete) [inline source/book/include/marlin/book/ROOTAdapter.h]: `using RHist = ...` to cope with other typedefs

File: source/book/include/marlin/book/ROOTAdapter.h [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+      using RH3I = ROOT::Experimental::RH3I ;
+
+      /**
+       *	@brief Buffer size for Histograms used in Shared mode.
+       *	- larger →  less synchronisation points
+       *	- larger →  more memory consumption.
+       *	@note can set with the CMAKE  Variable \cod {NAME}
+       */
+      constexpr std::size_t HistogramFillerBufferSize
+#ifdef HISTOGRAM_FILLER_BUFFER_SIZE
```
Comment (rete) [inline source/book/include/marlin/book/ROOTAdapter.h]:
```cpp
#ifndef MARLIN_HIST_FILLER_BUFFER_SIZE
#define MARLIN_HIST_FILLER_BUFFER_SIZE 1024
#endif
constexpr std::size_t HistogramFillerBufferSize = MARLIN_HIST_FILLER_BUFFER_SIZE ;
```

File: source/book/include/marlin/book/ROOTAdapter.h [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        = 1024 ;
+#endif
+
+      /**
+       *	@brief produce concurrent filler for one Histogram.
+       *	@tparam T Histogram type
+       */
+      template < class T >
+      using RHistConcurrentFillManager
+        = ROOT::Experimental::RHistConcurrentFillManager< T, 1024 > ;
```
Comment (rete) [inline source/book/include/marlin/book/ROOTAdapter.h]: Use `HistogramFillerBufferSize` instead and after.

File: source/book/include/marlin/book/Selection.h [hunk 20]
```diff
@@ -0,0 +1,120 @@
...
+    // -- MarlinBook forward declaration
+    class BookStore ;
+
+    /**
+     *	@brief Contains references to entries.
+     *	Which satisfy the condition.
+     *	Used to doing action on a range of entries.
+     */
+    class Selection {
+      friend BookStore ;
```
Comment (rete) [inline source/book/include/marlin/book/Selection.h]: Can't we get rid of this friend declaration ? We can expose a bit more the API if needed

File: source/book/include/marlin/book/Selection.h [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+       *	@brief Construct Selection from range of Entries.
+       *	@param begin,end range of Entries
+       * 	@param cond Condition to filter Entries.
+       */
+      template < typename T >
+      static Selection find( T begin, T end, const Condition &cond ) ;
+
+    public:
+      /// type for iteration through a Selection.
+      using iterator = typename std::vector< Entry >::const_iterator ;
```
Comment (rete) [inline source/book/include/marlin/book/Selection.h]: call it `const_iterator` if it is a `const_iterator`

File: source/book/include/marlin/book/ROOTAdapter.h [hunk 22]
```diff
@@ -0,0 +1,80 @@
+#pragma once
+
+// -- std includes
+#include <type_traits>
+
+// -- ROOT includes
+#include "ROOT/RHist.hxx"
+#include "ROOT/RHistConcurrentFill.hxx"
+#include "ROOT/RHistData.hxx"
+#include "ROOT/span.hxx"
```
Comment (rete) [inline source/book/include/marlin/book/ROOTAdapter.h]: Include RSpan instead

File: source/book/include/marlin/book/ROOTAdapter.h [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+      /// Type for Describing a Histogram Axis.
+      using RAxisConfig = ROOT::Experimental::RAxisConfig ;
+
+      /// Generic Histogram.
+      template < int D, typename T, template < int, class > class... STAT >
+      using RH = ROOT::Experimental::RHist< D, T, STAT... > ;
+
+      /// Histogram merge function.
+      template < typename TO, typename FROM >
+      void ADD( const TO &to, const FROM &from ) {
```
Comment (rete) [inline source/book/include/marlin/book/ROOTAdapter.h]: `ADD` -> `addHists`

File: source/book/include/marlin/book/BookStore.h [hunk 24]
```diff
@@ -0,0 +1,155 @@
+#pragma once
+
+// -- std includes
+#include <memory>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+// -- map ROOT to MarlinBook types
+#include "marlin/book/ROOTAdapter.h"
```
Comment (rete) [inline source/book/include/marlin/book/BookStore.h]: Group the includes here

File: source/book/include/marlin/book/BookStore.h [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+
+    /**
+     *	@brief Base Class for Entry Data, for similar behavior.
+     */
+    template < class T >
+    class EntryDataBase {} ;
+    template <>
+    class EntryDataBase< void > {
+    public:
+      EntryDataBase()                        						= default ;
```
Comment (rete) [inline source/book/include/marlin/book/BookStore.h]: Fix the indent here

File: source/book/include/marlin/book/BookStore.h [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+       *	@throw std::out_of_range key not exist in Store.
+       */
+      Entry &get( const EntryKey &key ) { return _entries[key.hash]; }
+
+    public:
+      template < class T >
+      auto book( const std::string_view &path,
+                 const std::string_view &name,
+                 const T &               data ) {
+        // TODO static assert and type check
```
Comment (rete) [inline source/book/include/marlin/book/BookStore.h]: TODO ?

File: source/book/include/marlin/book/Condition.h [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+      /// copy constructor.
+      Condition( const Condition & ) = default ;
+
+      /**
+       *	@brief test Condition on key.
+       * 	@return true if key matches the condition.
+       */
+      bool operator()( const EntryKey &key ) const { return _fiterFn( key ); }
+
+      // TODO: 'and' and 'or' are keywords in c++(and = &&, or = ||)
```
Comment (rete) [inline source/book/include/marlin/book/Condition.h]: TODO ?

File: source/book/include/marlin/book/Entry.h [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+
+      /// access key data from entry.
+      const EntryKey &key() const { return _key; }
+
+      /**
+       *	@brief check if entry is valid.
+       *	check if there is a reference stored or not.
+       * 	@return true if entry is valid.
+       */
+      bool valid() const { return static_cast< bool >( _entry ); }
```
Comment (rete) [inline source/book/include/marlin/book/Entry.h]: Use nullptr comparison to make it explicit

File: source/book/include/marlin/book/EntryData.h [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+
+    /**
+     *	@brief Data selection to identify and manage an Entry.
+     */
+    struct EntryKey {
+
+      /// default constructor.
+      EntryKey() : type{std::type_index( typeid( void ) )} {}
+
+      // TODO: clean solution
```
Comment (rete) [inline source/book/include/marlin/book/EntryData.h]: TODO ?

File: source/book/include/marlin/book/Handle.h [hunk 30]
```diff
@@ -0,0 +1,60 @@
...
+    class BookStore ;
+    class MemLayout ;
+
+    /**
+     *	@brief class which basic functionality for every handle.
+     *	@tparam T type which should be handled.
+     */
+    template < typename T >
+    class BaseHandle {
+      std::shared_ptr< MemLayout > _mem ;
```
Comment (rete) [inline source/book/include/marlin/book/Handle.h]: Members documentation

File: source/book/include/marlin/book/Hist.h [hunk 31]
```diff
@@ -0,0 +1,465 @@
...
+
+    /// trait specialisation for Histograms.
+    template < int D, typename T, template < int, class > class... STAT >
+    struct trait< types::RH< D, T, STAT... > > {
+      /**
+       *	@brief merging two histograms using Add.
+       *	@note \code {RHist::GetEntries()} is broken after that.
+       */
+      static void
+      Merge( const std::shared_ptr< types::RH< D, T, STAT... > > &dst,
```
Comment (rete) [inline source/book/include/marlin/book/Hist.h]: `merge` to stick to conventions

File: source/book/src/BookStore.cc [hunk 32]
```diff
@@ -0,0 +1,187 @@
...
+    void BookStore::addEntry( const std::shared_ptr< EntryBase > &entry,
+                              EntryKey &                          key ) {
+      EntryKey k = key ;
+      k.hash     = _entries.size() ;
+      _entries.push_back( Entry( entry, k ) ) ;
+    }
+
+    //--------------------------------------------------------------------------
+
+    template < class T, typename... Args_t >
```
Comment (rete) [inline source/book/src/BookStore.cc]: All templates in header files

File: source/book/src/Hist.cc [hunk 33]
```diff
@@ -0,0 +1,274 @@
...
+// -- MarlinBook includes
+#include "marlin/book/Handle.h"
+#include "marlin/book/MemLayout.h"
+#include "marlin/book/ROOTAdapter.h"
+
+namespace marlin {
+  namespace book {
+    using namespace types ;
+
+    template < int D, typename T, template < int, class > class... STAT >
```
Comment (rete) [inline source/book/src/Hist.cc]: This should also go in the header file

File: source/book/src/Selection.cc [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+// -- MarlinBook includes
+#include "marlin/book/Condition.h"
+#include "marlin/book/MemLayout.h"
+
+namespace marlin {
+  namespace book {
+
+    Selection::Selection( const Selection &sel,
+                          const Condition &cond,
+                          ComposeStrategie strategie )
```
Comment (rete) [inline source/book/src/Selection.cc]: Strategie -> Strategy

File: source/include/marlin/Exceptions.h (hunk 4 above)
Comment (rete) [reply source/include/marlin/Exceptions.h:140]: We can iterate on this later on and make another pull request. Keep it for the time being

File: test/CMakeLists.txt [hunk 35]
```diff
@@ -17,6 +17,12 @@ if( MARLIN_LCIO )
   link_libraries( ${LCIO_LIBRARIES} )
 endif()
 
+if( MARLIN_BOOK )
+	#find_package(MarlinBook CONFIG REQUIRED )
+	include_directories( ${PROJECT_SOURCE_DIR}/source/book/include)
+	link_libraries( "${PROJECT_SOURCE_DIR}/build/lib/libMarlinBook.so" )
```
Comment (rete) [inline test/CMakeLists.txt]:
```cmake
link_libraries( MarlinBook )
```

File: test/tests/test-book-store.cc [hunk 36]
```diff
@@ -0,0 +1,223 @@
...
+#include "marlin/book/Selection.h"
+#include "ROOT/RHistData.hxx" 
+#include "ROOT/RHist.hxx"
+#include "marlin/book/ROOTAdapter.h"
+
+#include <memory>
+#include <string>
+
+template<typename T>
+void MergeaHist(const std::shared_ptr<T>& dst, const std::shared_ptr<T>& src) {
```
Comment (rete) [inline test/tests/test-book-store.cc]: Use your function implemented in the framework

File: test/tests/test-book-store.cc [hunk 37]
```diff
@@ -0,0 +1,223 @@
...
+  ROOT::Experimental::Add(*dst, *src);
+}
+
+std::string mergedUnicStr() {
+  static std::size_t num = 0;
+  return std::to_string(++num);
```
Comment (rete) [inline test/tests/test-book-store.cc:20]: WARNING: Fine here, but std::to_string is not thread-safe

File: test/tests/test-mem-layout.cc [hunk 38]
```diff
@@ -0,0 +1,75 @@
...
+		: bins{a0, a1, a2, a3, a4} {}
+};
+
+void MergeType1(const std::shared_ptr<Type1>& p1, const std::shared_ptr<Type1>& p2) {
+	for(int i = 0; i < 5; ++i) {
+		p1->bins[i] += p2->bins[i];
+	}
+}
+template<typename ... Args_t>
+class SharedType1 : public SharedMemLayout<Type1, MergeType1, Args_t ... > {
```
Comment (rete) [inline test/tests/test-mem-layout.cc:26]: Why to you need to sub-class here ?

File: test/tests/test-mem-layout.cc [hunk 39]
```diff
@@ -0,0 +1,75 @@
...
+public:
+	SharedType1 (std::size_t num_inst, Args_t ... args) : SharedMemLayout<Type1, MergeType1, Args_t ... >{num_inst, args ...}{}
+};
+
+void MergeHist(const std::shared_ptr<RH1I>& p1, const std::shared_ptr<RH1I>& p2) {
```
Comment (rete) [inline test/tests/test-mem-layout.cc]: Use your nice framework here

**Review by rete — CHANGES_REQUESTED** [review]

**Comment by rete** [comment]
General comments:
- Check your indent. It is still not correct in a few places
- The expression AMT doesn't look good. We need to sit down and find a new expression for it
- Check the naming conventions, in particular for upper/lower case function name

Else it looks good ! Nice job !

File: source/CMakeLists.txt (hunk 7 above)
Comment (JBenda) [reply source/CMakeLists.txt]: not any longer, CMakeLists.txt in `source/book/` should fix it

File: .gitignore (hunk 8 above)
Comment (JBenda) [reply .gitignore:13]: not on purpose, ... sorry :sweat:

File: CMakeLists.txt (hunk 9 above)
Comment (JBenda) [reply CMakeLists.txt]: sorry, debug line ...

File: source/book/include/marlin/book/BookStore.h (hunk 26 above)
Comment (JBenda) [reply source/book/include/marlin/book/BookStore.h]: Not any longer, because of the of the book function wich is called. Should be save.

File: source/book/include/marlin/book/Condition.h (hunk 27 above)
Comment (JBenda) [reply source/book/include/marlin/book/Condition.h]: more an discuss, is Not, And, Or ok for the functions? because it's not the naming convention. When we overload the `&&` `||` operator `condition.and(condition2)` is valid ...

File: source/book/include/marlin/book/Entry.h (hunk 11 above)
Comment (JBenda) [reply source/book/include/marlin/book/Entry.h]: I need a mapping from type to merge function ..., but, an overloaded function should do it also

File: source/book/include/marlin/book/Entry.h (hunk 12 above)
Comment (JBenda) [reply source/book/include/marlin/book/Entry.h]: also, question, 
Should we log it if the creation failed because of a type mismatch or invalid Flag?
because when it's failed it's an internal error

File: source/book/include/marlin/book/EntryData.h (hunk 29 above)
Comment (JBenda) [reply source/book/include/marlin/book/EntryData.h]: not any longer, sorry, but may ugly

File: source/book/include/marlin/book/Selection.h (hunk 20 above)
Comment (JBenda) [reply source/book/include/marlin/book/Selection.h]: we can, the only drawback the that the iteration constructer than is public, which should be no problem, because the definition is in the .cc file

File: source/book/src/BookStore.cc (hunk 32 above)
Comment (JBenda) [reply source/book/src/BookStore.cc]: this is only the definition, the declaration is in the header file [`BookStore.h`](https://github.com/JBenda/MarlinMT/blob/8c5e0a393c7b0c9532e3c98a87ec5eadb1ccf03e/source/book/include/marlin/book/BookStore.h#L94)

File: source/book/src/Hist.cc (hunk 33 above)
Comment (JBenda) [reply source/book/src/Hist.cc]: because constructor should be every time in the header ?, if yes i must change a few more things.

File: test/tests/test-book-store.cc (hunk 37 above)
Comment (JBenda) [reply test/tests/test-book-store.cc:20]: ok, thanks for the note

File: test/tests/test-mem-layout.cc (hunk 38 above)
Comment (JBenda) [reply test/tests/test-mem-layout.cc:26]: for binding the merge function with the class (for Type1), we don't really need it. But the notation, this way is more likely used when you used the MemLayout, because the alternative is to construct it with:
`SharedMemLayout<Type1, MergeType1, int, int, int, int, int> sMem(…)`

File: docs/user-manual/booking-system.md [hunk 40]
```diff
@@ -0,0 +1,144 @@
+The `BookStore` is an Object which managed Histograms for later universal access.
```
Comment (rete) [inline docs/user-manual/booking-system.md]: Can you move this file to a new directory `internal-documentation` instead of `user-manual` ?

**Review by rete — APPROVED** [review]
